pressbooks/pressbooks
# feat: allow filters for books api endpoint

Issue: https://github.com/pressbooks/pressbooks/issues/3312

This PR adds necessary changes to handle new API parameters for `books` endpoint. The parameters handled are:
- `license_code[]` (array) to filter books by multiple license codes. For example `Public Domain` or `CC BY`.
- `title[]` (array), to filter or exclude books by multiple titles. for example `LTI Demo` to get books containing `LTI Demo` string (insensitive case) or `-biology` to exclude books with `biology` within the title.
- `in_directory` (boolean) to filter books included or excluded from PB Directory. For example: `in_directory=true`.
- `words` (string) to filter books `>=` or `<=` to a certain number of words. For example: `words=gte_1000` will bring book with `>=1000` words, and `words=lte_60000` will bring books with `<= 60000` words.

### Testing case
Basic book API endpoint in your local: `https://pressbooks.test/wp/wp-json/pressbooks/v2/books`

Make sure you have 5 or more public books to consider basic cases.

`in_directory` parameter:
- Add some books to the catalog (Books list).
- In Network Level > Settings > Sharing & Privacy check the `Exclude non-catalogued public books from Pressbooks Directory.` option and save it.
- add `?in_directory=true` parameter to the API endpoint. You must see only books with `bookDirectoryExcluded: false` in your list of books.
- now, change it to `?in_directory=false` parameter to the API endpoint. You must see only books with `bookDirectoryExcluded: true` in your list of books.
- Now, in Network Level > Settings > Sharing & Privacy uncheck the `Exclude non-catalogued public books from Pressbooks Directory.` option and save it.
- Go to some of your books and some of them in Settings > Sharing and Privacy, set `Pressbooks Directory` option to `Yes, I want this book to be listed in Pressbooks directory`, and for some of them, set it to `No`.
- in the basic API book endpoint you should see some `bookDirectoryExcluded` field as `true` and others as `false`.
- Use `in_directory` parameter as true and then as false to ensure you get the expected results.

`license_code[]` parameters:
- Add different licenses to your books.
- Use `?license_code[]` parameter to filter them by license code. You could use multiple licenses like `?license_code[]=CC BY&license_code[]=Public Domain` for example.
- Make sure the licenses match with the parameters.

`titles[]` parameters:
- Filter books by multiple titles like `title[]=test&title[]=demo` to get books containing `test` or `demo` within the title.
- You can exclude some books by specifying strings you don't want to be contained within the title. For example `title[]=lti&title[]=-demo'

`words` parameter:
- Check the above description to see how you can use it.
